packages/fxa-auth-server/scripts/audit-tokens.ts

```diff
@@ -0,0 +1,519 @@
+  logResult(name, query, result);
+  emitStats(name, result);
+  return result;
+}
+
+/** Looks for devices that have been orphaned, or have a parent row which was orphaned. */
+export async function auditOrphanedDeviceRows() {
+  // Orphaned devices. A slightly more complex query that explicitly checks that
+  // neither devices > sessionTokens > account nor devices > refreshToken > client exist.
+  const query = `
+  SELECT
+    *,
+    100 * total_missing_both / total AS percent_missing_both,
+    100 * total_missing_refresh_token / total AS percent_missing_refresh_token,
+    100 * total_missing_session_token / total AS percent_missing_session_token
+  FROM
+  (
+    SELECT
+      COUNT(*) as total,
+      COUNT( IF (missing_session_token = 1 and missing_refresh_token = 1, 1, NULL)) AS total_missing_both,
+      COUNT( IF (missing_refresh_token = 1, 1, NULL)) AS total_missing_refresh_token,
+      COUNT( IF (missing_session_token = 1, 1, NULL)) AS total_missing_session_token
+    FROM
+    (
+      SELECT
+        d.id,
+        case
+          WHEN s.tokenId is NULL OR a.uid is NULL
+          THEN 1 ELSE 0 END AS missing_session_token,
+        case
+          WHEN r.token is NULL OR c.id is NULL
+          THEN 1 ELSE 0 END AS missing_refresh_token
+      FROM
+        (
+          SELECT id, sessionTokenId, refreshTokenId
+          FROM ${tables.devices}
+          ${buildLimit(tables.devices)}
+        ) as d
+        left join ${tables.sessionTokens} s on s.tokenId = d.sessionTokenId
+          left join ${tables.accounts} a on s.uid = a.uid
+        left join ${tables.refreshTokens} r on r.token = d.refreshTokenId
+          left join ${tables.clients} c on c.id = r.clientId
+    ) AS status
+  ) AS totals
+  `;
+
+  const name = `${tables.devices}.OrphanedRows.On-Many`;
+  const result = await audit(name, query);
+  decorateResultWithTableStats(tables.devices, result);
+  logResult(name, query, result);
+  emitStats(name, result);
+  return result;
+}
+
+/** Runs audits according to current cli arguments */
+async function auditAll() {
+  // We always audit row counts. These queries are fast, and
+  // row counts are used by subsequent queries.
+  for (const table of Object.values(tables)) {
+    await auditRowCounts(table);
+  }
+
+  // If requested audit the age distribution of rows in the table.
+  if (program.auditAge) {
+    let set: any[] = [
+      [tables.accountCustomers, 'createdAt', 'uid'],
+      [tables.accountResetTokens, 'createdAt', 'tokenId'],
+      [tables.accounts, 'createdAt', 'uid'],
+      [tables.devices, 'createdAt', 'id'],
+      [tables.emails, 'createdAt', 'id'],
+      [tables.keyFetchTokens, 'createdAt', 'tokenId'],
+      [tables.passwordChangeTokens, 'createdAt', 'tokenId'],
+      [tables.passwordForgotTokens, 'createdAt', 'tokenId'],
+      [tables.sentEmails, 'sentAt', 'id'],
+      [tables.sessionTokens, 'createdAt', 'tokenId'],
+      [tables.sessionTokens, 'lastAccessTime', 'tokenId'],
```

Comment:
We already know this to be unreliable at best; what are you hoping to find with it?